patches/server/0965-Make-banners-modifiable-with-both-BannerMeta-and-Blo.patch

```diff
@@ -0,0 +1,270 @@
+diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
+index f45e4acee69bd95ff2e41feaf44f5414d2e40df0..5d47d47358dfa082e9b7bf142a24bd2e3ac7dd24 100644
+--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
++++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
+@@ -362,7 +362,7 @@ public final class CraftItemStack extends ItemStack {
+             case WHITE_WALL_BANNER:
+             case YELLOW_BANNER:
+             case YELLOW_WALL_BANNER:
+-                return new CraftMetaBanner(item.getTag());
++                return new CraftMetaBanner(item.getTag(), material); // Paper - Provide the material
+             case ALLAY_SPAWN_EGG:
+             case AXOLOTL_SPAWN_EGG:
+             case BAT_SPAWN_EGG:
+diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBanner.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBanner.java
+index 881f221517dc062a34489df6f7e006af9540953f..63f0bf38ceb00541e0fd891cbf860f98efb7935f 100644
+--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBanner.java
++++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBanner.java
+@@ -18,7 +18,7 @@ import org.bukkit.craftbukkit.util.CraftMagicNumbers;
+ import org.bukkit.inventory.meta.BannerMeta;
+ 
+ @DelegateDeserialization(CraftMetaItem.SerializableMeta.class)
+-public class CraftMetaBanner extends CraftMetaItem implements BannerMeta {
++public class CraftMetaBanner extends CraftMetaBlockState implements BannerMeta { // Paper - Make banners also modifiable with BlockStateMeta
+ 
+     private static final Set<Material> BANNER_MATERIALS = Sets.newHashSet(
+             Material.BLACK_BANNER,
+@@ -59,14 +59,29 @@ public class CraftMetaBanner extends CraftMetaItem implements BannerMeta {
+     static final ItemMetaKey PATTERNS = new ItemMetaKey("Patterns", "patterns");
+     static final ItemMetaKey COLOR = new ItemMetaKey("Color", "color");
+     static final ItemMetaKey PATTERN = new ItemMetaKey("Pattern", "pattern");
++    // Paper start - Add banner materials to CraftMetaBlockState's material list
++    static {
++        CraftMetaBlockState.BLOCK_STATE_MATERIALS.addAll(CraftMetaBanner.BANNER_MATERIALS);
++    }
++    // Paper end - Add banner materials to CraftMetaBlockState's material list
+ 
+     private DyeColor base;
+     private List<Pattern> patterns = new ArrayList<Pattern>();
+-
++    // Paper start - Create new constructor to properly handle different banners
++    private final org.bukkit.block.Banner bannerBlock;
++    @Deprecated
+     CraftMetaBanner(CraftMetaItem meta) {
+-        super(meta);
++        this(meta, Material.WHITE_BANNER);
++
++        new org.bukkit.plugin.AuthorNagException("Usage of internal deprecated constructor in paper: CraftMetaBanner(CraftMetaItem), this will incorrectly assume the item is a white banner").printStackTrace();
++    }
```

Comment:
We can just remove this constructor imo. It's not part of the api (obv) and its not even public.